crates/ruff_linter/src/rules/pylint/rules/boolean_chained_comparison.rs

```diff
@@ -0,0 +1,145 @@
+
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Contains chained boolean comparison that can be simplified")
+    }
+
+    fn fix_title(&self) -> Option<String> {
+        Some("Simplify chained boolean comparisons".to_string())
+    }
+}
+
+/// PLR1716
+pub(crate) fn boolean_chained_comparison(checker: &mut Checker, expr_bool_op: &ExprBoolOp) {
+    // early out for boolean expression without multiple compare values
+    if expr_bool_op.values.len() == 1 {
+        return;
+    }
+
+    // early out for non `and` boolean operations
+    if expr_bool_op.op != BoolOp::And {
+        return;
+    }
+
+    // early exit when not all expressions are compare expressions
+    // TODO: check if this can even happen?
+    if !expr_bool_op.values.iter().all(Expr::is_compare_expr) {
+        return;
+    }
+
+    // retrieve all compare statements from expression
+    let compare_exprs: Vec<&ExprCompare> = expr_bool_op
+        .values
+        .iter()
+        .map(|stmt| stmt.as_compare_expr().unwrap())
+        .collect();
+
+    let results: Vec<Result<(), BooleanChainedComparison>> = compare_exprs
+        .iter()
+        .tuple_windows::<(&&ExprCompare, &&ExprCompare)>()
+        .filter(|(left_compare, right_compare)| {
+            are_compare_expr_simplifiable(left_compare, right_compare)
+        })
+        .map(|(left_compare, right_compare)| {
+            let Expr::Name(left_compare_right) = left_compare.comparators.first().unwrap() else {
+                return Ok(());
+            };
+
+            let Expr::Name(ref right_compare_left) = right_compare.left.as_ref() else {
+                return Ok(());
+            };
+
+            if left_compare_right.id() != right_compare_left.id() {
+                return Ok(());
+            }
+
+            Err(BooleanChainedComparison {
+                variable: left_compare_right.id().clone(),
+                range: TextRange::new(left_compare.range.start(), right_compare.range.end()),
+                replace_range: TextRange::new(
+                    left_compare_right.range.start(),
+                    right_compare_left.range.end(),
+                ),
+            })
+        })
+        .collect();
+
+    let results: Vec<BooleanChainedComparison> = results
+        .into_iter()
+        .filter(Result::is_err)
+        .map(|result| result.err().unwrap())
+        .collect();
+
+    checker.diagnostics.extend(
+        results
+            .into_iter()
+            .map(|result| {
+                let variable = result.variable.clone();
+                let range = result.range;
+                let replace_range = result.replace_range;
+                let mut diagnostic = Diagnostic::new(result, range);
+                diagnostic.set_fix(Fix::safe_edit(Edit::range_replacement(
+                    variable.to_string(),
+                    replace_range,
+                )));
```

Comment:
Nit: I'm inclined to inline those variables as they're only used once (simplifies reading because there's less aliasing that my mind has to keep track of)

```suggestion
                let mut diagnostic = Diagnostic::new(result, result.range);
                diagnostic.set_fix(Fix::safe_edit(Edit::range_replacement(
                    variable.to_string(),
                    result.replace_range,
                )));
```

Reply:
I was able to move the creation of the edit up like this:

```rust
            let range = result.range;
            let edit = Edit::range_replacement(result.variable.to_string(), result.replace_range);
            let mut diagnostic = Diagnostic::new(result, range);
                        diagnostic.set_fix(Fix::safe_edit(edit));
```

Preventing the `let variable = result.variable.clone();` and `let replace_range = result.replace_range;`.

However, when trying to do the suggested `Diagnostic::new(result, result.range);` I get borrow checker errors:

```rust
            let edit = Edit::range_replacement(result.variable.to_string(), result.replace_range);
            let mut diagnostic = Diagnostic::new(result, result.range);
            diagnostic.set_fix(Fix::safe_edit(edit));
```

Gives me these errors:

```rust
error[E0382]: use of moved value: `result`
   --> crates/ruff_linter/src/rules/pylint/rules/boolean_chained_comparison.rs:104:58
    |
102 |         .extend(results.into_iter().map(|result| {
    |                                          ------ move occurs because `result` has type `rules::pylint::rules::boolean_chained_comparison::BooleanChainedComparison`, which does not implement the `Copy` trait
103 |             let edit = Edit::range_replacement(result.variable.to_string(), result.replace_range);
104 |             let mut diagnostic = Diagnostic::new(result, result.range);
    |                                                  ------  ^^^^^^^^^^^^ value used here after move
    |                                                  |
    |                                                  value moved here
```

What do you think would be the best thing to do here?

Reply:
Oh, I didn't notice that `result` is used in `Diagnostic::new`. Assigning a temporary variable is then indeed necessary to satisfy the borrow checker.